### PR TITLE
fix(fqdn): hostname named instance #21

### PR DESCRIPTION
**Pull request by alexirion10**
DNS / fqdn name lookup doesn't work on Named Instances

#21

### CODE REVIEW

**Review by Kaneraz — COMMENTED** [review]
Have you tested this with an SSL connection per the original issue in #21?

**Comment by alexirion10** [comment]
I have not.  Do you have any guides for that?
`Dns.GetHostEntry` throws an exception on named instances

**Comment by Kaneraz** [comment]
It's in the connection string, I think encrypt=yes, but we should have it set in the consuming code.

File: src/SmoFacade/Listener.cs [hunk 1]
```diff
@@ -128,11 +129,21 @@ private static string AgListenerName(string dataSource)
...
-    private static Server AgListenerNameToServer(ref SqlConnectionStringBuilder connBuilder, string agInstanceName,
-      string credentailName)
+    private static Server AgListenerNameToServer(ref SqlConnectionStringBuilder connBuilder, string agInstanceName)
     {
-      connBuilder.DataSource = Dns.GetHostEntry(agInstanceName).HostName;
-      return new Server(connBuilder.ToString(), credentailName);
+      try {
+        connBuilder.DataSource = Dns.GetHostEntry(agInstanceName).HostName;
+      }
+      catch(Exception) {
```
Comment (Kaneraz) [inline src/SmoFacade/Listener.cs:137]: Is there a more specific exception we should catch here?

[Probably irrelevant, see below]

File: src/SmoFacade/Listener.cs [hunk 2]
```diff
@@ -128,11 +129,21 @@ private static string AgListenerName(string dataSource)
...
+        // NamedInstances: chop instance name, resolve DNS, slap instance name back on!
+        var parts = agInstanceName.Split('\\');
```
Comment (Kaneraz) [inline src/SmoFacade/Listener.cs:139]: We should just do this from the get go to determine if it's a named instance. 

What are the limitations on SQL hostnames and instance names?
https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms143531(v=sql.105)?redirectedfrom=MSDN
https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/naming-conventions-for-computer-domain-site-ou#:~:text=The%20maximum%20length%20of%20the,from%20the%20NETBIOS%20host%20name.

They can't have a backslash in the name, so we should be good start with this.

**Review by Kaneraz — CHANGES_REQUESTED** [review]
Looks like this resolves the potential issues with certificate errors. We can probably clean this up a bit more to avoid using exceptions for control flow.

File: src/SmoFacade/Listener.cs (hunk 2 above)
Comment (alexirion10) [reply src/SmoFacade/Listener.cs:139]: will do.
I figured FDS names wouldn't have a `\`, but its reassuring that msft restricts it for all named instances

File: src/SmoFacade/Listener.cs (hunk 1 above)
Comment (alexirion10) [reply src/SmoFacade/Listener.cs:137]: SocketException
https://docs.microsoft.com/en-us/dotnet/api/system.net.dns.gethostname?view=netcore-3.1

File: src/SmoFacade/Listener.cs (hunk 2 above)
Comment (alexirion10) [reply src/SmoFacade/Listener.cs:139]: @Kaneraz - The problem with this actually is that `\` is not restricted from a servername, only a Named instance.
If I `Split` on backslash and it's a servername, what would happen?  

The exception control flow works nicely to only split the named instance if DNS doesn't match the first attempt

Comment (Kaneraz) [reply src/SmoFacade/Listener.cs:139]: I read it the other way, that our servername can't have a `\` in it.

Comment (alexirion10) [reply src/SmoFacade/Listener.cs:139]: gotcha, I guess neither can have `\` in the name

**Review by Kaneraz — APPROVED** [review]
Looks great, thanks!

**Pull request merged**